rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,272 @@
+ - `touch_policy=cached` + `TouchPolicy=never` = failure
+
+#### Role Options
+
+To make yubikey login required or optional on a per-role basis, you can set the same options in a Role.
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: role-name
+spec:
+  yubikey_login: 
+    required: true
+    pin_policy: once
+    touch_policy: cached
+```
+
+#### Combined yubikey requirements
+
+Each user's yubikey requirements will be determined by:
+ 1. Combining their role options to form the strictest setting
+ 2. Combining the strictest role options settings with the cluster auth preference, with role options taking precedence (if set).
+
+Note: A user's yubikey requirement settings cannot be easily determined before generating the yubikey private key and locking in its pin/touch policy settings, which occurs before logging in. Thus, it is the responsibility of the user to figure out their own pin/touch policy settings before logging in, or else they will get `access denied` errors.
+
+#### Examples:
+
+Example 1:
+ - All users must use yubikey login
+ - Users with the `admin` role must also use `pin=always` and `touch=always` (to secure adminstrative infrastructure)
+ - Users with the `superadmin` role are not required to use yubikey login (superadmin roles are inherently unsafe, might as well own it)
+```yaml
+kind: cluster_auth_preference
+version: v2
+metadata:
+  name: cluster-auth-preference
+spec:
+  yubikey_login: 
+    required: true
+---
+kind: role
+version: v5
+metadata:
+  name: admin
+spec:
+  yubikey_login: 
+    pin_policy: always
+    touch_policy: always
+---
+kind: role
+version: v5
+metadata:
+  name: superadmin
+spec:
+  yubikey_login: 
+    required: false
+```
+
+Example 2:
+ - Users with `yubikey-pin` role must use yubikey login with `pin=always` and `touch=cached|always`
+ - Users with `yubikey-touch` role must use yubikey login with `pin=once|always` and `touch=always`
+ - Users with both the `yubikey-pin` and `yubikey-touch` roles must use yubikey login  with `pin=always` and `touch=always`
+ - Users with neither the `yubikey-pin` nor `yubikey-touch` roles are not required to use yubikey login
+```yaml
+kind: cluster_auth_preference
+version: v2
+metadata:
+  name: cluster-auth-preference
+spec:
+  yubikey_login: 
+    required: false
+    pin_policy: once
+    touch_policy: cached
+---
+kind: role
+version: v5
+metadata:
+  name: yubikey-pin
+spec:
+  yubikey_login: 
+    required: true
+    pin_policy: always
+---
+kind: role
+version: v5
+metadata:
+  name: yubikey-touch
+spec:
+  yubikey_login: 
+    required: true
+    pin_policy: never
+```
+
+### UX
+
+#### yubikey slots
+
+Yubikey 5 has 5 different [key slots](https://docs.yubico.com/hardware/yubikey/yk-5/tech-manual/yk5-apps.html#slot-information) to choose from, one of which is reserved for attestation. `Slot 9a: PIV Authentication` is the most fitting option, and is the same slot used by `yubikey-agent`.
```

Comment:
How hard would it be to let the user choose a difference slot?

Reply:
Not hard, we could give the user a flag like `--yubikep-piv-slot` which defaults to `9a` but can be set to any of the 4 valid slots.

Reply:
+1